### ISSUE TITLE
Subscription messages can silently fail to post back to client

### ISSUE THREAD

**Issue by braudabaugh** [graphql-dotnet/server#250]
When generating subscription messages in rapid succession, some messages may never make it back to the client.  I've debugged  the issue to the Subscription.OnNext function calling _writer.Post here:

https://github.com/graphql-dotnet/server/blob/a41d7c318760226af17ed33b9e4ebe8e900032a7/src/Transports.Subscriptions.Abstractions/Subscription.cs#L62

The call to _writer.Post returns false when the messages are lost.  A false return value is meant to indicate that the Post could not be completed without blocking, but the Subscription.OnNext function does not handle this case.

My solution is to use _write.SendAsync instead, as it will send the messages even if it must block.  Here are my implementation of Subscription.OnNext:

```
        public void OnNext(ExecutionResult value)
        {
            _logger.LogDebug("Subscription: {subscriptionId} got data", Id);
            var task = Task.Run(() => _writer.SendAsync(new OperationMessage
            {
                Type = MessageType.GQL_DATA,
                Id = Id,
                Payload = value
            }));
            task.Wait();
        }
```

I am not sure if this is the best solution, but it is working for my use case.

**Comment by sungam3r** [comment]
Why to wrap `SendAsync` in `Task.Run` and then wait for it? Why not just use
```c#
public void OnNext(ExecutionResult value)
        {
            _logger.LogDebug("Subscription: {subscriptionId} got data", Id);
            _writer.SendAsync(new OperationMessage
            {
                Type = MessageType.GQL_DATA,
                Id = Id,
                Payload = value
            }).GetAwaiter().GetResult();
        }
```

Also note that `SendAsync` returns `Task`, **but** underlying `ITargetBlock.SendAsync` returns `Task<bool>` - https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.dataflowblock.sendasync?f1url=https%3A%2F%2Fmsdn.microsoft.com%2Fquery%2Fdev15.query%3FappId%3DDev15IDEF1%26l%3DEN-US%26k%3Dk(System.Threading.Tasks.Dataflow.DataflowBlock.SendAsync%60%601);k(DevLang-csharp)%26rd%3Dtrue&view=netcore-2.2

**Comment by braudabaugh** [comment]
Using the Task.Run technique avoids causing a possible deadlock by allocating a new thread to do the actual blocking.  See this reference:

https://stackoverflow.com/questions/17248680/await-works-but-calling-task-result-hangs-deadlocks/32429753#32429753

Since the OnNext method did not currently cause the calling thread to block, I was hesitant to change that behavior.  Since I'm not super familiar with the graphql-dotnet code base, I'll defer to others to determine if blocking the calling thread is okay or not.

**Comment by sungam3r** [comment]
`Task.Run` does not necessary allocate new thread. In any case you introduce blocking on calling thread which is not good.

**Comment by braudabaugh** [comment]
You're right.  Unless Task.Wait was removed which would cause another set of problems.
Perhaps there is a solution lower down in the code.

**Comment by jinhong-** [comment]
Any updates regarding this issue? We are currently facing the same problem as well. Messages sent in quick successions causes missing messages sent to the client

**Comment by sungam3r** [comment]
I did not intend to study this problem in the near future. I'm not well prepared for the topic of subscriptions yet.

**Comment by braudabaugh** [comment]
I settled on the following change in my fork based on conversations with @sungam3r:
```
        public void OnNext(ExecutionResult value)
        {
            _logger.LogDebug("Subscription: {subscriptionId} got data", Id);
            var result = _writer.Post(new OperationMessage
            {
                Type = MessageType.GQL_DATA,
                Id = Id,
                Payload = value
            });
            if (!result)
            {
                _writer.SendAsync(new OperationMessage
                {
                    Type = MessageType.GQL_DATA,
                    Id = Id,
                    Payload = value
                }).GetAwaiter().GetResult();
            }
        }
```
It seems to be working, however some of the other open issues related to subscriptions, #255 for example, are preventing us from doing much with this functionality.

**Comment by jinhong-** [comment]
What do you think about this solution? I was thinking, instead of handling the message sending as a subscriber, why not use the reactive extensions to extend the observable pipeline instead. This way i was able to use the async APIs without using any synchronous task calls:
```
    public class Subscription
    {
        private readonly Action<Subscription> _completed;
        private readonly ILogger<Subscription> _logger;
        private readonly IWriterPipeline _writer;
        private IDisposable _unsubscribe;

        public Subscription(string id,
            OperationMessagePayload payload,
            SubscriptionExecutionResult result,
            IWriterPipeline writer,
            Action<Subscription> completed,
            ILogger<Subscription> logger)
        {
            _writer = writer;
            _completed = completed;
            _logger = logger;
            Id = id;
            OriginalPayload = payload;

            Subscribe(result);
        }

        public string Id { get; }

        public OperationMessagePayload OriginalPayload { get; }


        private async Task SendComplete()
        {
            _logger.LogDebug("Subscription: {subscriptionId} completing", Id);
            await _writer.SendAsync(new OperationMessage
            {
                Type = MessageType.GQL_COMPLETE,
                Id = Id
            });

            _completed?.Invoke(this);
            _unsubscribe?.Dispose();
        }

        private async Task SendData(ExecutionResult value)
        {
            _logger.LogDebug("Subscription: {subscriptionId} got data", Id);
            await _writer.SendAsync(new OperationMessage
            {
                Type = MessageType.GQL_DATA,
                Id = Id,
                Payload = value
            });
        }

        public Task UnsubscribeAsync()
        {
            _logger.LogDebug("Subscription: {subscriptionId} unsubscribing", Id);
            _unsubscribe.Dispose();
            return _writer.SendAsync(new OperationMessage
            {
                Type = MessageType.GQL_COMPLETE,
                Id = Id
            });
        }

        private void Subscribe(SubscriptionExecutionResult result)
        {
            var stream = result.Streams.Values.Single();
            _unsubscribe = stream.Synchronize()
                .SelectMany(value => Observable.FromAsync(() => SendData(value)))
                .Concat(Observable.FromAsync(SendComplete))
                .Subscribe();

            _logger.LogDebug("Subscription: {subscriptionId} subscribed", Id);
        }
    }
```

**Comment by jinhong-** [comment]
I raised PR #279 with the new code. Tested it in my project and it seems to work

**Comment by jinhong-** [comment]
@sungam3r any chance to get my PR merged into the main code for the next release?

**Comment by danmalcolm** [comment]
We are seeing the same issue in a project. We've tested with a build based on PR #279 which resolves the problem for us.

This is a good PR. I'd really like to see this merged in.

**Comment by sungam3r** [comment]
I'd really like to see this merged in **too** but I just can't afford to merge code that I don't understand. For now, the subscription mechanism is out of my attention.

**Comment by jinhong-** [comment]
Is there anything I can do to help to increase the confidence if having this PR merged?

**Comment by sungam3r** [comment]
Unlikely, but I still hope that I will have time to look and understand the subscription code. Or you can turn to another maintainer who understands the issue better and who can confidently merge this code.

**Comment by BenjaBobs** [comment]
An alternative route could be to solved https://github.com/graphql-dotnet/server/issues/289.

**Comment by jinhong-** [comment]
> An alternative route could be to solved #289. This would take the concern of managing websockets and dataflow out of managing subscriptions, which could make the subscription code easier to understand. Would require a re-write though.

This looks like using SignalR as the transport layer. Does that mean that it will no longer conform to the apollo sbscriptions specs?

**Comment by BenjaBobs** [comment]
Not necessarily. It would mean replacing the `Transports.Subscriptions.Abstractions` and `Transports.Subscriptions.WebSockets` code with similar [SignalR code](https://docs.microsoft.com/en-us/aspnet/core/tutorials/signalr?tabs=visual-studio&view=aspnetcore-3.1).

@pekkah did a [SignalR implementation](https://github.com/pekkah/tanka-graphql/tree/master/src/graphql.server/WebSockets) for his Tanka-GraphQL project.

**Comment by pekkah** [comment]
Nowadays TGQL supports both. The additional burden of SignalR is that it requires client side library also to make it work with Apollo.

**Comment by DavidKiff** [comment]
We are hitting this today too... surely this is fundamental to streaming data?  Is this fixed in v3?

**Comment by jinhong-** [comment]
@DavidKiff Unless there was some fundamental change on handling failed sending of messages in the DataFlow block, i don't think so.. I can see that my PR #279 with the potential fix is still pending, with some merge conflicts now. What we did was to recompile our own version of the library with the fix and use that instead

**Comment by cfahrenholz** [comment]
We are also facing this issue, some messages are silently dropped. Unfortunately we recognized this almost right before our release. I am very suprised that this basic feature "subscriptions" is not ready for production.
@joemcbride can you please provide a statement if you plan to fix this or to merge PR #279 ? Any chance to get this into 2.3.x (asp net core 2.1)?

**Comment by Dravere** [comment]
I think I just hit this one as well. And with a really simple use-case. All I'm doing is basically calling `OnNext(data)` and then immediately afterwards `OnCompleted()` on my observers.

```
observer.OnNext(data);
observer.OnCompleted();
```

This will drop the `complete` message on the websocket connection.

I'm really unfamiliar with DataFlow. But looking at the code, could this be related to the `BoundedCapacity` of 1 in the message writer of the `WebSocketWriterPipeline`?
https://github.com/graphql-dotnet/server/blob/b028d3d07ffdfbdcd80364c692d13b2a128bc8c9/src/Transports.Subscriptions.WebSockets/WebSocketWriterPipeline.cs#L46

Why does this have to be one?

**Comment by Shane32** [comment]
That complete sense now.  The default value is -1 (infinite).  Let me see if I can reproduce the bug, hopefully within a reproducible test, and see if changing the value fixes the problem.

**Comment by Shane32** [comment]
@Dravere The setting you refer to was removed in #199 .  Which version of the nuget package are you using?

**Comment by Shane32** [comment]
@cfahrenholz Which version nuget package are you using?  The 3.5 alpha nuget packages all contain a potential fix via #199 .

**Comment by Dravere** [comment]
@Shane32 I'm on the latest stable version (3.4.0). Apparently this fix is in the alpha version 3.5.0 that has been in development for over a year. Could we perhaps get a version 3.4.1 or will 3.5.0 finally release anytime soonish?

**Comment by Shane32** [comment]
I have tested the sample provided by @cfahrenholz with the commit immediately prior and subsequent to #199 and have confirmed that #199 does indeed fix this issue.

@Dravere 3.5 is based on GraphQL 3.0, which should be released within a couple days (maybe today).  This project and other dependent libraries will follow shortly thereafter.

@pekkah Can answer if we can release a 3.4.1.  I've prepped a pull request.

**Comment by Dravere** [comment]
@Shane32 Well that is amazing news. I probably can postpone subscriptions a bit and might directly jump to GraphQL.Net 3.0. Glad to hear that this release is coming soon as well. :+1:

**Comment by cfahrenholz** [comment]
> 
> 
> @cfahrenholz Which version nuget package are you using? The 3.5 alpha nuget packages all contain a potential fix via #199 .

@Shane32 we use version 3.4.0. In the decompiler I see the BoundedCapacity = 1.

Since we currently cannot switch to .net core > 2.1 we would need this fix in 3.4.1.

Thank you @Shane32!

**Comment by sungam3r** [comment]
3.5.0 is netstandard2.0 compatible. https://www.nuget.org/packages/GraphQL.Server.Transports.AspNetCore/3.5.0-alpha0073

**Comment by Dravere** [comment]
Finally was able to make the switch. Works perfectly fine in GraphQL Server 4.0.1. Thank you!

**Comment by Shane32** [comment]
> @Shane32 I'm on the latest stable version (3.4.0). Apparently this fix is in the alpha version 3.5.0 that has been in development for over a year. Could we perhaps get a version 3.4.1 or will 3.5.0 finally release anytime soonish?

3.4.1 has been published with the necessary fix (#199) included.  Sorry it took so long, but glad to hear you've been able to make the switch to 4.0.1.
